### PR TITLE
Add include and exclude attributes to EnvironmentTagHelper

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #5671 

cc @Eilon

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 1]
```diff
@@ -30,13 +34,33 @@ public EnvironmentTagHelper(IHostingEnvironment hostingEnvironment)
 
         /// <summary>
         /// A comma separated list of environment names in which the content should be rendered.
+        /// If the current environment is also in the <see cref="Exclude"/>  list, the content will not be rendered.
         /// </summary>
         /// <remarks>
         /// The specified environment names are compared case insensitively to the current value of
         /// <see cref="IHostingEnvironment.EnvironmentName"/>.
         /// </remarks>
         public string Names { get; set; }
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: @DamianEdwards @Eilon, do we plan to remove this in the future? If so should we add a remark saying this is deprecated?

Comment (DamianEdwards) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Nah, I'd just leave it. It's basically an alias for `Include`

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 2]
```diff
@@ -11,10 +11,14 @@ namespace Microsoft.AspNetCore.Mvc.TagHelpers
...
     /// <see cref="ITagHelper"/> implementation targeting &lt;environment&gt; elements that conditionally renders
     /// content based on the current value of <see cref="IHostingEnvironment.EnvironmentName"/>.
+    /// If the environment is not listed in explicitly identified <see cref="Names"/> or <see cref="Include"/> lists or
+    /// if it is in the <see cref="Exclude"/> list, the content will not be rendered. 
     /// </summary>
     public class EnvironmentTagHelper : TagHelper
     {
         private static readonly char[] NameSeparator = new[] { ',' };
+        private static readonly char[] IncludeEnvironmentsSeparator = new[] { ',' };
+        private static readonly char[] ExcludeEnvironmentsSeparator = new[] { ',' };
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Why imply we'd ever use different separators for `Include` and `Exclude`? Can use `NameSeparator` everywhere.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 3]
```diff
@@ -68,27 +92,75 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
                 // environments matched the current environment. Suppress the output in this case.
                 output.SuppressOutput();
+                return;
+            }
+
+            if (hasExcludeEnvironments)
+            {
+                // No matching exclude environment name found, do nothing
+                return;
             }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Why does this block exist? Method is about to return anyhow.

Also do not need the `return` at [line 157](https://github.com/aspnet/Mvc/pull/5722/files#diff-226e99a1a5f76aa4841ff3004e7dd7ffR157) or the `hasExcludeEnvironments` variable.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 4]
```diff
@@ -68,27 +92,75 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+                        if (environment.Equals(currentEnvironmentName))
+                        {
+                            output.SuppressOutput();
+                            return;
+                        }
+                    }
+                }
+            }
+
             var hasEnvironments = false;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Just need one `hasIncludeEnvironments` variable for both `Names` and `Includes`.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs [hunk 5]
```diff
@@ -66,6 +66,97 @@ public void ShowsContentWhenCurrentEnvironmentIsNotSet(string namesAttribute, st
...
+        [InlineData("Development", "", "", "Development")]
+        [InlineData("", "Development", "", "Development")]
+        [InlineData("", "Development", "Test", "Development")]
+        [InlineData("Development", "", "Test", "Development")]
+        [InlineData("Development", "Development", "", "Development")]
+        [InlineData("Test", "Development", "", "Development")]
+        [InlineData("Development", "Test", "", "Development")]
+        [InlineData("Development", "Development", "Test", "Development")]
+        [InlineData("Development", "Test", "Test", "Development")]
+        [InlineData("Test", "Development", "Test", "Development")]
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: Test `null` as well as empty strings. Same for the data added below. (Note: You don't really need to test every positive `""`, `"Development"`, `"Test"` combination.)

Also suggest removing the `environmentName` parameter since its value never varies.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs [hunk 6]
```diff
@@ -66,6 +66,97 @@ public void ShowsContentWhenCurrentEnvironmentIsNotSet(string namesAttribute, st
...
+            var context = MakeTagHelperContext(
+                attributes: new TagHelperAttributeList {
+                    { "names", namesAttribute },
+                    { "include", includeAttribute },
+                    { "exclude", excludeAttribute },
+                });
+            var output = MakeTagHelperOutput("environment", childContent: content);
+            var hostingEnvironment = new Mock<IHostingEnvironment>();
+            hostingEnvironment.SetupProperty(h => h.EnvironmentName);
+            hostingEnvironment.Object.EnvironmentName = environmentName;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: Why is this line needed?

**Comment by dougbu** [comment]
Overall comment: Need a couple of tests covering comma-separated values for `Include` and `Exclude.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 7]
```diff
@@ -11,10 +11,14 @@ namespace Microsoft.AspNetCore.Mvc.TagHelpers
     /// <summary>
     /// <see cref="ITagHelper"/> implementation targeting &lt;environment&gt; elements that conditionally renders
     /// content based on the current value of <see cref="IHostingEnvironment.EnvironmentName"/>.
+    /// If the environment is not listed in explicitly identified <see cref="Names"/> or <see cref="Include"/> lists or
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: `If the environment is not listed in <see cref="Names"/> or <see cref="Include"/>, or if it is in <see cref="Exclude"/>, the content will not be rendered.`

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 8]
```diff
@@ -30,13 +34,33 @@ public EnvironmentTagHelper(IHostingEnvironment hostingEnvironment)
 
         /// <summary>
         /// A comma separated list of environment names in which the content should be rendered.
+        /// If the current environment is also in the <see cref="Exclude"/>  list, the content will not be rendered.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: you have an extra space after `<see cref="Exclude"/>`.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: What if we made it a true alias for Include? Aka, have its getter/setter point to `Include`. That way we could add a doc comment stating it was an alias which would move people towards Include and avoid any confusion in the future.

It'd also simplify the code in process a decent amount.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 9]
```diff
@@ -11,10 +11,14 @@ namespace Microsoft.AspNetCore.Mvc.TagHelpers
     /// <summary>
     /// <see cref="ITagHelper"/> implementation targeting &lt;environment&gt; elements that conditionally renders
     /// content based on the current value of <see cref="IHostingEnvironment.EnvironmentName"/>.
+    /// If the environment is not listed in explicitly identified <see cref="Names"/> or <see cref="Include"/> lists or
+    /// if it is in the <see cref="Exclude"/> list, the content will not be rendered. 
     /// </summary>
     public class EnvironmentTagHelper : TagHelper
     {
         private static readonly char[] NameSeparator = new[] { ',' };
+        private static readonly char[] IncludeEnvironmentsSeparator = new[] { ',' };
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Just realized @dougbu already left this feedback 🙈

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: That still won't restrict people from using both at the same time and that will cause some confusion.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 10]
```diff
@@ -68,27 +92,75 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+            if (Exclude != null)
+            {
+                var tokenizer = new StringTokenizer(Exclude, ExcludeEnvironmentsSeparator);
+                foreach (var item in tokenizer)
+                {
+                    var environment = item.Trim();
+                    if (environment.HasValue && environment.Length > 0)
+                    {
+                        hasExcludeEnvironments = true;
+                        if (environment.Equals(currentEnvironmentName))
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: StringComparison.OrdinalIgnoreCase

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 11]
```diff
@@ -68,27 +92,75 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
             {
-                // This instance had at least one non-empty environment specified but none of these
+                var tokenizer = new StringTokenizer(Include, IncludeEnvironmentsSeparator);
+                foreach (var item in tokenizer)
+                {
+                    var environment = item.Trim();
+                    if (environment.HasValue && environment.Length > 0)
+                    {
+                        hasIncludeEnvironments = true;
+                        if (environment.Equals(currentEnvironmentName))
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: StringComparison.OrdinalIgnoreCase

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs (hunk 5 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: Also include a few cases to ensure that the comparison is case insensitive

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs (hunk 6 above)
Comment (jbagga) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: If the `EnvironmentName` is not set, it's `null` and always returns from the [null check](https://github.com/aspnet/Mvc/pull/5722/files#diff-226e99a1a5f76aa4841ff3004e7dd7ffR89) for `currentEnvironmentName`

Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: Something is busted with mocking if you need both of these lines. The "real" `EnvironmentName` property should never be called.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Yeah keep both properties. It's a bit gross, but the alternatives seem worse.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs (hunk 6 above)
Comment (jbagga) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: `hostingEnvironment.SetupProperty(h => h.EnvironmentName, "Development");` seems to work. Should I modify all the older tests to do the same?

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 12]
```diff
@@ -30,13 +34,33 @@ public EnvironmentTagHelper(IHostingEnvironment hostingEnvironment)
...
+        /// If the current environment is also in the <see cref="Exclude"/> list, the content will not be rendered.
+        /// </summary>
+        /// <remarks>
+        /// The specified environment names are compared case insensitively to the current value of
+        /// <see cref="IHostingEnvironment.EnvironmentName"/>.
+        /// </remarks>
+        public string Include { get; set; }
+
+        /// <summary>
+        /// A comma separated list of environment names in which the content should not be rendered.
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Replace `should` with `will`. In fact, make this change in all the comments (assuming it applies 😄 ).

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 3 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: I partially agree with this comment. I think the return at line 157 is important because it ensures the method will end right there and then. It will prevent a future change later in this method from inadvertently running.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: Yes please. Much less head-scratching for all.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 12 above)
Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: It applies for `Exclude`. But for `Include` and `Names`, if the environment is also in `Exclude`, it will not be rendered. Leaving those as "should"

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 7 above)
Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: This is a little tricky. If the user does not specify `Names`, `Include` or `Exclude`, the content is rendered for all environments. But if they do specify `Names` and/or `Include`, but the current environment is not in it, then it is suppressed.  

Maybe
`If the environment is not listed in the specified <see cref="Names"/> or <see cref="Include"/>, or if it is in <see cref="Exclude"/>, the content will not be rendered.`

**Comment by jbagga** [comment]
@dougbu @Eilon  @ajaybhargavb Updated

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs [hunk 13]
```diff
@@ -66,6 +66,95 @@ public void ShowsContentWhenCurrentEnvironmentIsNotSet(string namesAttribute, st
             ShouldShowContent(namesAttribute, environmentName);
         }
 
+
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/EnvironmentTagHelperTest.cs]: nit: Remove extra line

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs [hunk 14]
```diff
@@ -68,27 +90,67 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+                            return;
+                        }
                     }
                 }
             }
 
             if (hasEnvironments)
             {
-                // This instance had at least one non-empty environment specified but none of these
+                // This instance had at least one non-empty environment (name or include) specified but none of these
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: name => names

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs (hunk 12 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs]: Makes sense, thanks!
